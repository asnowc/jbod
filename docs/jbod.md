## JS JBOD

### Base data type:

0000_xxxx :

| DEC | BIN  | type        | content length                              |
| --- | ---- | ----------- | ------------------------------------------- |
| 0   | 0000 | void        | 0 (Represents termination in map and array) |
| 1   | 0001 | null        | 0                                           |
| 2   | 0010 | undefined   | 0                                           |
| 3   | 0011 | true        | 0                                           |
| 4   | 0100 | false       | 0                                           |
| 5   | 0101 | int         | 4                                           |
| 6   | 0110 | bigint      | 8                                           |
| 7   | 0111 | double      | 8                                           |
| 8   | 1000 | \*objectId  | dynamicLenData                              |
| 9   | 1001 | arrayBuffer | contentLen(dynamicLenData) + content        |
| 10  | 1010 | \*string    | arrayBuffer                                 |
| 11  | 1011 | \*regExp    | string                                      |
| 12  | 1100 | function    | string[]&string //暂未实现                  |
| 13  | 1101 | array       | n                                           |
| 14  | 1110 | object      | n                                           |
| 15  | 1111 | \*symbol    |                                             |

0001_xxxx:

| DEC | BIN  | type    | content length |
| --- | ---- | ------- | -------------- |
| 16  | 0000 | \*error | object         |
| 17  | 0001 | map     |                |
| 18  | 0010 | set     | array          |
| 19  | 0011 |         |                |
| 20  | 0100 |         |                |
| 21  | 0101 |         |                |
| 22  | 0110 |         |                |
| 23  | 0111 |         |                |
| 24  | 1000 |         |                |
| 25  | 1001 |         |                |
| 26  | 1010 |         |                |
| 27  | 1011 |         |                |
| 28  | 1100 |         |                |
| 29  | 1101 |         |                |
| 30  | 1110 |         |                |
| 31  | 1111 |         |                |

#### JBOD Array:

```
|--1-|---n---|  |--1-|---n---| ... |--1-|---n---| |--1-|
|type|content|  |type|content| ... |type|content| |void|

```

#### JBOD Map:

```
|--1-|<dynamicContent>|---n---|  |--1-|<dynamicContent>|---n---| ... |--1-|<dynamicContent>|---n---| |--1-|
|type|      key       |content|  |type|      key       |content| ... |type|      key       |content| |void|

```

#### dynamicLenData

| byte | max               | real    | content                          |
| ---- | ----------------- | ------- | -------------------------------- |
| 1    | 0x7f              | 7bit    | 0xxxxxxx                         |
| 2    | 0x3fff            | 6bit 1B | 10xxxxxx ff                      |
| 3    | 0x1fffff          | 5bit 2B | 110xxxxx ff ff                   |
| 4    | 0xfffffff(255MB)  | 4bit 3B | 1110xxxx ff ff ff                |
| 5    | 0x7_ffffffff      | 3bit 4B | 11110xxx ff ff ff ff             |
| 6    | 0x3ff_ffffffff    | 2bit 5B | 111110xx ff ff ff ff ff          |
| 7    | 0x1ffff_ffffffff  | 1bit 6B | 1111110x ff ff ff ff ff ff       |
|      |                   |         |                                  |
| 8    | 0xffffff_ffffffff | 7B      | 11111110 ff ff ff ff ff ff ff    |
| 9    |                   | 8B      | 11111111 ff ff ff ff ff ff ff ff |

number(47 bit): 0~512 TB - 1
id(7 bytes): 0~65535 TB -1

#### dynamicContent

```
<dynamicLenData> content
```
