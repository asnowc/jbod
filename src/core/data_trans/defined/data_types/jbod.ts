import type { TypeDataWriter, DataWriter, EncodeContext } from "../type.js";

export class JbodWriter implements TypeDataWriter {
  constructor(data: any, ctx: EncodeContext) {
    this.type = ctx.toTypeCode(data);
    this.writer = new ctx[this.type](data, ctx);
    this.byteLength = this.writer.byteLength + 1;
  }
  private writer: DataWriter;
  readonly type: number;
  byteLength: number;
  encodeTo(buf: Uint8Array, offset: number): number {
    buf[offset] = this.type;
    return this.writer.encodeTo(buf, offset + 1);
  }
}
