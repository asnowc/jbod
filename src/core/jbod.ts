import { JbodTrans } from "./data_trans/trans.js";
export { StructTrans } from "./data_trans/struct.js";
export { JbodTrans, type JbodTransConfig } from "./data_trans/trans.js";
class UserJbodTrans extends JbodTrans {
  /**
   * @public
   * @remarks 将数据编码为携带类型的 Uint8Array, 这会比 encodeContent 多一个字节
   */
  encode(data: any) {
    const res = this.createWriter(data);
    const buf = new Uint8Array(res.byteLength);
    res.encodeTo(buf, 0);
    return buf;
  }
  /**
   * @public
   * @remarks 将数据编码为不携带类型的 Uint8Array, 这会比 encode 少一个字节
   */
  encodeContent(data: any) {
    const res = this.createContentWriter(data);
    const buf = new Uint8Array(res.byteLength);
    res.encodeTo(buf, 0);
    return buf;
  }
}
export default new UserJbodTrans();
export * from "./data_trans/mod.js";
