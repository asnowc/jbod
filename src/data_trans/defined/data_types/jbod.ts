import type { TypeDataWriter, DataWriter, DecodeContext, EncodeContext } from "../type.ts";
import type { DecodeResult } from "../../../type.ts";

import { DataType, UnsupportedDataTypeError } from "../const.ts";
import { NO_CONTENT } from "../data_types/fixed_len.ts";
import * as numberTrans from "./fixed_len.ts";
import { decodeU32D, decodeU64D, zigzagDecodeI64, zigzagDecodeI32 } from "../../../varints/mod.ts";

export class JbodWriter implements TypeDataWriter {
  constructor(data: any, ctx: EncodeContext) {
    this.type = ctx.toTypeCode(data);
    this.writer = new ctx[this.type](data, ctx);
    this.byteLength = this.writer.byteLength + 1;
  }
  private writer: DataWriter;
  readonly type: number;
  byteLength: number;
  encodeTo(buf: Uint8Array, offset: number): number {
    buf[offset] = this.type;
    return this.writer.encodeTo(buf, offset + 1);
  }
}
/**
 * @remarks 获取数据类型对应的代码
 */
export function toTypeCode(this: EncodeContext, data: any): number {
  let type: number;
  switch (typeof data) {
    case "number":
      if (data % 1 !== 0 || data < -2147483648 || data > 2147483647) type = DataType.f64;
      else type = DataType.dyI32;
      break;
    case "bigint":
      type = DataType.dyI64;
      break;
    case "boolean":
      return data ? DataType.true : DataType.false;
    case "string":
      type = DataType.string;
      break;
    case "undefined":
      return DataType.undefined;
    case "symbol":
      type = DataType.symbol;
      break;
    case "object":
      if (data !== null) return getClassTypeCode(data, this.classTypes);
    default:
      //TODO: function 待实现
      return DataType.null;
  }
  return type;
}
function getClassTypeCode(data: object, classTypes: Map<object, number>) {
  let constructor = Reflect.getPrototypeOf(data);
  let type: number | undefined;
  while (constructor) {
    type = classTypes.get(constructor as any);
    if (type !== undefined) return type;
    constructor = Reflect.getPrototypeOf(constructor);
  }
  return DataType.anyRecord;
}
const { encoder: F64Writer, decoder: decodeF64 } = numberTrans.f64;
const { decoder: decodeI32 } = numberTrans.i32;
const { decoder: decodeI64 } = numberTrans.i64;

const f64 = F64Writer.prototype.encodeTo;
/**
 * @remarks 判断数据类型，并返回对应 带类型的 DataWriter
 */
export function fastJbodWriter(data: any, ctx: EncodeContext): TypeDataWriter {
  let type: number;
  switch (typeof data) {
    case "number":
      if (data % 1 !== 0 || data < -2147483648 || data > 2147483647)
        return { byteLength: 8, encodeTo: f64, data, type: DataType.f64 } as any;
      else type = DataType.dyI32;
      break;
    case "bigint":
      type = DataType.dyI64;
      break;
    case "boolean":
      return { type: data ? DataType.true : DataType.false, byteLength: 0, encodeTo: NO_CONTENT.encodeTo };
    case "string":
      type = DataType.string;
      break;
    case "undefined":
      return { type: DataType.undefined, byteLength: 0, encodeTo: NO_CONTENT.encodeTo };
    case "symbol":
      type = DataType.symbol;
      break;
    case "object": {
      if (data === null) return { type: DataType.null, byteLength: 0, encodeTo: NO_CONTENT.encodeTo };
      type = getClassTypeCode(data, ctx.classTypes);
      break;
    }
    default:
      return { type: DataType.null, byteLength: 0, encodeTo: NO_CONTENT.encodeTo };
  }
  const writer = new ctx[type](data, ctx) as any;
  writer.type = type;
  return writer;
}

export function fastDecodeJbod(buf: Uint8Array, offset: number, ctx: DecodeContext, type: number): DecodeResult<any> {
  switch (type) {
    case DataType.i32:
      return decodeI32(buf, offset, ctx);
    case DataType.i64:
      return decodeI64(buf, offset, ctx);
    case DataType.f64:
      return decodeF64(buf, offset, ctx);
    case DataType.dyI32: {
      const res = decodeU32D(buf, offset) as { value: number; byte: number };
      return { data: zigzagDecodeI32(res.value), offset: offset + res.byte };
    }
    case DataType.dyI64: {
      const res = decodeU64D(buf, offset);
      return { data: zigzagDecodeI64(res.value), offset: offset + res.byte };
    }
    case DataType.true:
      return { data: true, offset };
    case DataType.false:
      return { data: false, offset };
    case DataType.null:
      return { data: null, offset };
    case DataType.undefined:
      return { data: undefined, offset };
    default: {
      const decode = ctx[type];
      if (typeof decode !== "function") throw new UnsupportedDataTypeError(DataType[type] ?? type);
      return decode(buf, offset, ctx);
    }
  }
}
export function jbodDecoder(buf: Uint8Array, offset: number, ctx: DecodeContext) {
  let type = buf[offset++];
  return fastDecodeJbod(buf, offset, ctx, type);
}
