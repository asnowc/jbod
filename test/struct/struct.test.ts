import { StructCodec, Struct } from "jbod";
import { describe, expect, test } from "vitest";
import { formatBin } from "../utils/mod.ts";

describe("编解码", function () {
  test("基础", function () {
    const s1 = StructCodec.define<{ f1: boolean; f2: number; f3: any }>({
      f1: { id: 1, type: "bool" },
      f2: { id: 2, type: "i32" },
      f3: 3,
    });
    const data = { f1: true, f2: 9, f3: "ab" };

    const u8Arr = s1.encode(data);
    expect(s1.decode(u8Arr)).toEqual({ data, offset: u8Arr.byteLength });
    expect(formatBin(u8Arr), "decode").toMatchSnapshot(); // 01(03) 02(00000009) 03(0a026162) 00
  });
  test("sample repeat", function () {
    const struct = StructCodec.define({
      f1: { id: 1, repeat: true, type: "dyI32" },
      f2: { id: 2, repeat: true },
    });
    const raw = { f1: [1, 2, 3], f2: [1, true] };
    const bin = struct.encode(raw);

    expect(struct.decode(bin)).toEqual({ data: raw, offset: bin.byteLength });
    expect(formatBin(bin), "decode").toMatchSnapshot(); // 01(03020406) 02(02070203) 00
  });
  test("struct repeat", function () {
    const arrItem = {
      k2: {
        id: 1,
        type: "dyI32",
      },
    } satisfies Struct;
    const struct = StructCodec.define({
      f1: {
        id: 4,
        repeat: true,
        type: arrItem,
      },
    });
    const raw = { f1: [{ k2: 10 }, { k2: 6 }] };
    const bin = struct.encode(raw);

    /* 
    f1: 04(02 011400 010c00) 00
    f1[0]: 01 14 00
    f1[2]: 01 0c 00
    */
    expect(formatBin(bin), "decode").toMatchSnapshot();
    expect(struct.decode(bin)).toEqual({ data: raw, offset: bin.byteLength });
    expect(formatBin(bin), "decode").toMatchSnapshot();
  });

  test("boolean", function () {
    const optionalStruct = StructCodec.define<Partial<{ f1: boolean; f2: boolean }>>({
      f1: { id: 1, type: "bool" },
      f2: { id: 2, type: "bool" },
    });
    const data = { f1: false, f2: true };
    const u8Arr = optionalStruct.encode(data);
    expect(optionalStruct.decode(u8Arr)).toEqual({ data, offset: u8Arr.byteLength });

    expect(formatBin(u8Arr)).toMatchSnapshot();
  });
  test("嵌套", function () {
    const optionalStruct = StructCodec.define<Partial<{ f1: number; f2: { c1: number; c2: number } }>>({
      f1: 1,
      f2: { id: 2, type: { c1: 1, c2: 2 } },
    });
    const raw = { f1: 8, f2: { c1: 9, c2: -1 } };
    const u8Arr = optionalStruct.encode(raw);
    expect(optionalStruct.decode(u8Arr)).toEqual({ data: raw, offset: u8Arr.byteLength });

    expect(formatBin(u8Arr)).toMatchSnapshot();
  });
});
describe("可选类型", function () {
  test("可选类型编解码", function () {
    const struct = StructCodec.define<Partial<{ f1: boolean; f2: number; f3: any }>>({
      f1: { id: 1, type: "bool", optional: true },
      f2: { id: 2, type: "i32" },
    });
    const raw = { f2: 9 };
    const u8Arr = struct.encode(raw);
    expect(struct.decode(u8Arr)).toEqual({ data: raw, offset: u8Arr.byteLength });
    expect(formatBin(u8Arr)).toMatchSnapshot();
  });
  test("字段缺失", function () {
    const data = { f1: 2 };
    const s1 = StructCodec.define({ f1: { id: 1, type: "dyI32" }, f2: { id: 2, type: "dyI32" } });
    expect(() => s1.encode(data), "默认不可选").toThrowError();

    const s2 = StructCodec.define({ f1: 1, f2: 2 });
    expect(() => s2.encode(data), "默认不可选").toThrowError();

    const s3 = StructCodec.define({ f1: { id: 1, type: { f2: 2 } } });
    expect(() => s3.encode({ f1: {} }), "默认不可选").toThrowError();
  });
  test("配置默认可选", function () {
    const data = { f1: 2 };

    const s2 = StructCodec.define({ f1: 1, f2: 2 }, { defaultOptional: true });
    expect(() => s2.encode(data), "默认不可选").not.toThrowError();
  });
});
