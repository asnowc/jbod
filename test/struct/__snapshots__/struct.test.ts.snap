// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`encode > boolean 1`] = `"0104020300"`;

exports[`encode > sample repeat 1`] = `"0103020406020207020300"`;

exports[`encode > struct repeat 1`] = `"0402011400010c0000"`;

exports[`encode > struct repeat 2`] = `"0402011400010c0000"`;

exports[`encode > 可选类型 1`] = `"020000000900"`;

exports[`encode > 基础 1`] = `"01030200000009030a02616200"`;

exports[`encode > 嵌套 1`] = `"010710020107120207010000"`;

exports[`可选类型 > 可选类型编解码 1`] = `"020000000900"`;

exports[`编解码 > boolean 1`] = `"0104020300"`;

exports[`编解码 > sample repeat 1`] = `"0103020406020207020300"`;

exports[`编解码 > struct repeat 1`] = `"0402011400010c0000"`;

exports[`编解码 > struct repeat 2`] = `"0402011400010c0000"`;

exports[`编解码 > 基础 1`] = `"01030200000009030a02616200"`;

exports[`编解码 > 嵌套 1`] = `"010710020107120207010000"`;
